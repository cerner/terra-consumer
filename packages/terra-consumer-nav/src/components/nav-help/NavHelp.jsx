import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import Button from 'terra-button'
import Arrange from 'terra-arrange';
import Popup from 'terra-popup';
import Modal from 'terra-modal';
import HelpToggleItem from './HelpToggleItem';
import OutlineInfo from '../../icons/OutlineInfo';
import styles from './NavHelp.scss';

const cx = classNames.bind(styles);

const propTypes = {
  help: PropTypes.arrayOf(PropTypes.shape({
    slug: PropTypes.string,
    nav_type: PropTypes.oneOf(['GROUPING', 'EXTERNAL', 'APPLICATION', 'MODAL']),
    target: PropTypes.string,
    text: PropTypes.string.isRequired,
    uri: PropTypes.string.isRequired,
    icon: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.element,
      ]),
    children: PropTypes.arrayOf(PropTypes.shape({
      slug: PropTypes.string,
      nav_type: PropTypes.oneOf(['GROUPING', 'EXTERNAL', 'APPLICATION', 'MODAL']),
      text: PropTypes.string.isRequired,
      uri: PropTypes.string.isRequired,
      icon: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.element,
      ]),
      children: PropTypes.array,
    })),
  })),
};

const defaultProps ={
  help:[{}],
};

const translations = {
  help: 'Help'
};

class NavHelp extends React.Component {
  constructor() {
    super();

    this.state = {
      isOpen: false,
    };

    this.openPopup = this.openPopup.bind(this);
    this.closePopup = this.closePopup.bind(this);
  }

  openPopup() {
    this.setState({isOpen: true});
  }

  closePopup() {
    this.setState({isOpen: false});
  }

  render() {
    const { help,...customProps} = this.props;
    
    const helpButton = (<Button
      id='nav-help-button' 
      onClick={this.openPopup}
      className={cx('nav-help')}
      icon={<svg className={cx('help-icon')} style={{paddingRight: '6px'}}><OutlineInfo /></svg>}
      text={translations.help}
      />);

    
    const popupContent = help.map((navItem,i) =>
      <div key={i} className={cx('divider-top')}>
       { navItem.children.length === 0 &&
         <a href={navItem.uri} onClick={()=> window.location = navItem.uri }  className={cx('nav-help-item')} style={{padding:'0px'}}>
           <Arrange
             className={cx('nav-help-item')}
             align="stretch"
             fitStart={<div className={cx('help-icon')}>{navItem.icon}</div>}
             fill={<div style={{paddingLeft:'10px'}}>{navItem.text}</div>}
            />
          </a>
       }
       { navItem.children.length > 0 && 
         <HelpToggleItem
           navItem={navItem.text}
           navSubItems={navItem.children}
           navIcon={navItem.icon}
         />      
       }
       </div>);

    return(
       <div {...customProps}>
         {helpButton}
         <Popup
          isOpen={this.state.isOpen}
          onRequestClose={this.closePopup}
          targetRef={() => document.getElementById('nav-help-button')}
          isArrowDisplayed
          contentWidth="320"
          contentHeight="640"
          contentAttachment='top left'
         >
         {<div>{popupContent}</div>}
         </Popup>
       </div>
      );
  }

}

NavHelp.propTypes = propTypes;
NavHelp.defaultProps = defaultProps;
export default NavHelp;

