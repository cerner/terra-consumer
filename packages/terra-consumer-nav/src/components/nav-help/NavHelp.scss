@import '../../variables';

:local {
  .nav-help {
    background-color: var(--terra-consumer--nav-help-button-background, #fff);
    border: 0;
    border-radius: 20px;
    box-shadow: 0 1px 1px #666;
    color: var(--terra-consumer--nav-help-button-color, #3d3d3d);
    display: inline-block;
    font-family: inherit;
    font-size: 16px;
    line-height: 20px;
    padding: 6px 13px;

    &:hover {
      background: var(--terra-consumer--nav-help-button-hover-background, #eee);
      cursor: pointer;
    }

    &:focus {
      background: var(--terra-consumer--nav-help-button-focus-background, #eee);
    }

    &:active {
      color: inherit;
    }
  }

  .help-item {
    background: var(--terra-consumer--nav-help-item-background, #fff);
    border: 0;
    border-bottom: 1px solid var(--terra-consumer--nav-help-item-border-color, #dedfe0);
    border-radius: 0;
    display: block;
    font-family: inherit;
    padding: 20px;
    text-decoration: none;
    width: 100%;

    &:hover {
      background: var(--terra-consumer--nav-help-item-hover-background, #eee);
      cursor: pointer;
    }

    &:focus {
      background: var(--terra-consumer--nav-help-item-focus-background, #eee);
    }
  }

  .help-item-text {
    color: var(--terra-consumer--nav-help-item-text-color, #383f42);
    font-size: 14px;
    line-height: 18px;
    text-align: left;
  }

  .help-subitem {
    color: var(--terra-consumer--nav-help-subitem-text-color, #64696c);
    font-size: 14px;
    line-height: 18px;
    text-decoration: none;
  }

  .icon-text-padding {
    padding-right: 10px;
  }

  .toggler-padding {
    padding: 0 !important;
  }

  .toggler-content-alignment {
    padding-right: 20px;
    padding-top: 20px;
    text-align: left;
  }

  .button-text-padding {
    padding-left: 6px;
  }

  .link-text-style {
    text-decoration: none;
  }

  .popup-content {
    border-radius: 5px;
    box-shadow: 0 1px 1px #666;
    margin: 0 15px 15px 0;
    overflow: hidden !important;
    position: relative;
  }
}
