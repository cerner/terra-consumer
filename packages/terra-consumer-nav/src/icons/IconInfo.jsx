/* eslint-disable */
import React from 'react';
import IconBase from 'terra-icon/lib/IconBase';

const SvgIcon = (customProps) => {
  const attributes = Object.assign({}, customProps);

  return (
    <IconBase {...attributes}>
      <path d="M24,1C11.317,1,1,11.318,1,24s10.317,23,23,23s23-10.318,23-23S36.683,1,24,1z M24,45C12.421,45,3,35.58,3,24S12.421,3,24,3s21,9.42,21,21S35.579,45,24,45z" />
      <circle cx="24" cy="36" r="2" />
      <path
        d="M24,11c-2.642,0-5.375,1.868-6.5,4.442c-0.222,0.506,0.01,1.096,0.516,1.317c0.508,0.222,1.097-0.01,1.316-0.516
        C20.141,14.395,22.147,13,24,13c2.663,0,5,2.336,5,5s-2.337,5-5,5c-0.553,0-1,0.448-1,1v5c0,0.552,0.447,1,1,1s1-0.448,1-1v-4.075
        c3.341-0.502,6-3.47,6-6.925C31,14.206,27.794,11,24,11z"
      />
    </IconBase>
  );
};

SvgIcon.displayName = "IconInfo";
SvgIcon.defaultProps = {"viewBox":"0 0 48 48","xmlns":"http://www.w3.org/2000/svg","data-name":"Layer 1"};

export default SvgIcon;
/* eslint-enable */
