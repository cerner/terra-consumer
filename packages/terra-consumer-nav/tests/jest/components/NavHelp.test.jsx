import React from 'react';
import NavHelp from '../../../src/components/nav-help/NavHelp';
import IconInfo from '../../../src/icons/IconInfo';

const helpItems = [
  {
    text: 'Technical Questions',
    uri: 'http://localhost:8083/',
    icon: (<IconInfo />),
    children: [{
      text: 'Need help using this portal or need to report an issue? Contact the support team at 123-xxx-xxxx',
      uri: 'http://localhost:8083/',
    }],
  },
  {
    text: 'Get Support ID',
    uri: 'http://localhost:8083/',
    icon: (<IconInfo />),
    children: [],
  },
];

describe('NavHelp button with pop/modal', () => {
  // Snapshot Tests
  it('should render a button with IconInfo,label and a pop/modal not opened', () => {
    const wrapper = shallow(<NavHelp help={helpItems} />);
    expect(wrapper).toMatchSnapshot();
  });
});
