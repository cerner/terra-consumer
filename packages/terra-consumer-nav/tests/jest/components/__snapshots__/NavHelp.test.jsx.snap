// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavHelp button with pop/modal should render a button with IconInfo,label and a pop/modal not opened 1`] = `
<IntlProvider
  locale="en-US"
  messages={
    Object {
      "nav_help": "Help",
      "nav_profile_title": "Settings",
      "nav_signout": "Sign Out",
    }
  }
>
  <div>
    <InjectIntl(NavHelp)
      help={
        Array [
          Object {
            "children": Array [
              Object {
                "text": "Need help using this portal or need to report an issue? Contact the support team at 123-xxx-xxxx",
                "uri": "http://localhost:8080/",
              },
            ],
            "icon": <IconOutlineQuestionMark
              height={16}
              viewBox="0 0 48 48"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
        />,
            "text": "Technical Questions",
            "uri": "http://localhost:8080/",
          },
          Object {
            "children": Array [],
            "icon": <IconOutlineQuestionMark
              height={16}
              viewBox="0 0 48 48"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
        />,
            "text": "Get Support ID",
            "uri": "http://localhost:8080/",
          },
        ]
      }
      id="nav-help-button"
    />
  </div>
</IntlProvider>
`;
